File: solutions/scala/luhn/src/main/scala/Luhn.scala
object Luhn:
  def valid(numberToValidate: String): Boolean =
    val number = numberToValidate.replace(" ", "")
    if !number.matches("\\d{2,}") then false
    else
      number.map(_.asDigit).reverse.zipWithIndex.map { case (digit, index) =>
        if index % 2 == 0 then digit
        else
          val x = digit * 2
          if x > 9 then x - 9 else x
      }.sum % 10 == 0

File: solutions/python/luhn/luhn.py
class Luhn:
    def __init__(self, card_num):
        self.card_num = card_num.replace(' ', '')

    def valid(self):
        return self.card_num.isdigit() and len(self.card_num) > 1 and sum(
            doubled_digit - 9 if doubled_digit > 9 else doubled_digit
            for doubled_digit in (
                digit * (1 + index % 2)
                for index, digit in enumerate(map(int, reversed(self.card_num)))
            )
        ) % 10 == 0
